nifi-extension-bundles/nifi-network-bundle/nifi-network-processors/src/main/java/org/apache/nifi/processors/network/pcap/Packet.java

```diff
@@ -0,0 +1,157 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package org.apache.nifi.processors.network.pcap;
+
+public class Packet {
+    static final int PACKET_HEADER_LENGTH = 16;
+    private ByteBufferReader io;
+    private long tsSec;
+    private long tsUsec;
+    private long inclLen;
+    private long origLen;
+    private long expectedLength;
+    private int totalLength;
+    private PCAP root;
+    private byte[] rawBody;
+    private String invalidityReason;
+
+    public Packet(ByteBufferReader io, PCAP root) {
+        this.root = root;
+        this.io = io;
+        read();
+    }
+
+    public Packet(byte[] headerArray, PCAP root) {
+        this.root = root;
+        this.io = new ByteBufferReader(headerArray);
+        read();
+    }
+
+    public Packet(long tSSec, long tSUsec, long inclLen, long origLen, byte[] rawBody) {
+        // packet header properties
+        this.tsSec = tSSec;
+        this.tsUsec = tSUsec;
+        this.inclLen = inclLen;
+        this.origLen = origLen;
+
+        // packet calculated properties
+        this.expectedLength = inclLen;
+        this.totalLength = PACKET_HEADER_LENGTH + rawBody.length;
+        this.rawBody = rawBody;
+        this.setValidity();
+    }
+
+    private void read() {
+        this.tsSec = this.io.readU4();
+        this.tsUsec = this.io.readU4();
+        this.inclLen = this.io.readU4();
+        this.origLen = this.io.readU4();
+
+        this.expectedLength = Math.min(inclLen(), root().getHeader().snaplen());
+
+        if (!this.io.isEof() && this.io.bytesLeft() >= expectedLength) {
+            this.rawBody = this.io.readBytes(expectedLength);
+        } else {
+            this.rawBody = new byte[0];
+        }
+        this.setValidity();
+        this.totalLength = PACKET_HEADER_LENGTH + this.rawBody.length;
+    }
+
+    private void setValidity() {
+        this.invalidityReason = null;
+        if (this.rawBody.length == 0) {
+            this.invalidityReason = "Packet body is empty";
+            return;
+        }
+        if (this.inclLen > this.origLen) {
+            this.invalidityReason = ("The reported length of this packet exceeds the reported length of the original packet "
+                + "as sent on the network; (" + this.inclLen + " > " + this.origLen + ")");
+            return;
+        }
+        if (this.origLen == 0) {
+            this.invalidityReason = "The reported original length of this packet as send on the network is 0.";
+            return;
+        }
+        if (this.inclLen == 0) {
+            this.invalidityReason = "The reported length of this packet is 0.";
+            return;
+        }
```

Comment:
Recommend reworking this to use `else if` instead of multiple short-circuit returns.